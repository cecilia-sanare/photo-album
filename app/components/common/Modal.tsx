import classNames from 'classnames';
import * as styles from './Modal.module.scss';
import { ReactNode, useEffect } from 'react';
import { Backdrop } from './Backdrop';
import { useCachedState } from '../../hooks/use-cached-state';

type ModalProps = {
  opened?: boolean;
  children: ReactNode;
  onClose?: () => void;
}

export function Modal({
  opened = false,
  children,
  onClose
}: ModalProps) {
  const [internallyOpened, setInternallyOpened] = useCachedState<boolean>(() => opened, [opened]);

  useEffect(() => {
    if (internallyOpened) {
      document.body.classList.add('no-scroll');
    } else {
      document.body.classList.remove('no-scroll');
    }
  }, [internallyOpened]);

  return (
    <>
      <Backdrop
        opened={internallyOpened}
        onClose={() => {
          setInternallyOpened(false);
          onClose?.();
        }}
      />
      <div
        className={classNames(
          styles.modal,
          internallyOpened && styles.opened
        )}
        data-testid='modal'
      >
        {children}
      </div>
    </>
  )
}
