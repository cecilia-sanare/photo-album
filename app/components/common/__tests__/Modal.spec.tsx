import { fireEvent, render } from '@testing-library/react';
import { Modal } from '../Modal';
import Chance from 'chance';

const chance = new Chance();

describe('<Modal />', () => {
  it('should render the Modal', () => {
    const component = render(<Modal>Content</Modal>);

    expect(component.baseElement).toMatchSnapshot();
  });

  describe('prop(opened)', () => {
    it('should open the modal', () => {
      const component = render(<Modal opened>Content</Modal>);

      expect(component.getByTestId('modal')).toHaveClass('opened');
    });

    it('should close the modal', () => {
      const component = render(<Modal>Content</Modal>);

      expect(component.getByTestId('modal')).not.toHaveClass('opened');
    });
  });

  describe('prop(onClose)', () => {
    it('should close the modal when the backdrop is clicked', async () => {
      const component = render(<Modal opened>Content</Modal>);

      await fireEvent.click(component.getByTestId('backdrop'));

      expect(component.getByTestId('modal')).not.toHaveClass('opened');
    });

    it('should fire off the onClose event when the backdrop is clicked', async () => {
      const onClose = jest.fn();
      const component = render(<Modal opened onClose={onClose}>Content</Modal>);

      expect(onClose).toHaveBeenCalledTimes(0);

      await fireEvent.click(component.getByTestId('backdrop'));

      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });
});
