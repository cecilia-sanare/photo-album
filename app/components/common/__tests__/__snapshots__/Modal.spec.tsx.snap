// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render the Modal 1`] = `
<body>
  <div>
    <div
      class="backdrop"
      data-testid="backdrop"
    />
    <div
      class="modal"
      data-testid="modal"
    >
      Content
    </div>
  </div>
</body>
`;
